x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -968,8 +973,28 @@ void populateAndRead(
                     );
 
                     if (gaps.isEmpty() == false) {
-                        for (SparseFileTracker.Gap gap : gaps) {
-                            executor.execute(fillGapRunnable(gap, writer, refs.acquireListener()));
+                        final SourceInputStreamFactory streamFactory = writer.sharedInputStreamFactory(gaps);
+                        logger.trace(
+                            () -> Strings.format(
+                                "fill gaps %s %s shared input stream factory",
+                                gaps,
+                                (streamFactory == null ? "without" : "with"),
+                                (streamFactory == null ? "" : " " + streamFactory)
+                            )
+                        );
+                        if (streamFactory == null) {
+                            for (SparseFileTracker.Gap gap : gaps) {
+                                executor.execute(fillGapRunnable(gap, writer, null, refs.acquireListener()));
+                            }
+                        } else {
+                            final List<AbstractRunnable> gapFillingTasks = gaps.stream()
+                                .map(gap -> fillGapRunnable(gap, writer, streamFactory, refs.acquireListener()))
+                                .toList();
+                            executor.execute(() -> {
+                                try (streamFactory) {
+                                    gapFillingTasks.forEach(Runnable::run);
```

Comment:
I think we should add a comment explaining that if a gap fails for whatever reason, the writer for following gaps will still be executed.

Reply:
Good point! See 07e488a86c990f7e844e6088840297b1b61ab2c9